### PR TITLE
Typings improvement

### PR DESCRIPTION
**Pull request by epessina**
Generic typings improvement. It includes the usage of generics to specify:
- type of service config
- route interfaces in handlers and decorators
- service responses

#### Checklist

- [x] your branch will not cause merge conflict with `master`
- [x] run `npm test`
- [x] tests are included
- [x] the documentation is updated or integrated accordingly with your changes
- [x] the code will follow the lint rules and style of the project
- [x] you are not committing extraneous files or sensible data

### CODE REVIEW

File: index.d.ts [hunk 1]
```diff
@@ -18,23 +18,24 @@ import * as fastify from 'fastify'
...
   function getServiceProxy(microserviceGatewayServiceName: string, options?: InitServiceOptions): Service
   interface environmentSchema {
     type: 'object',
-    required?: string[],
+    required?: string[] | readonly string[],
     properties: object
   }
 
-  type AsyncInitFunction = (service: DecoratedFastify) => Promise<void>
+  type AsyncInitFunction<ServiceConfig = NodeJS.Dict<string>> = (service: DecoratedFastify<ServiceConfig>) => Promise<void>
```
Comment (ilteoood) [inline index.d.ts:38]: The part ServiceConfig =... Is repeated multiple times.

Can you extract it in a common type?

File: index.d.ts [hunk 2]
```diff
@@ -57,18 +58,18 @@ declare namespace customPlugin {
...
   type RawCustomPluginMethod = 'GET' | 'POST' | 'PUT' | 'PATCH' | 'DELETE' | 'HEAD'
 
-  interface DecoratedFastify extends fastify.FastifyInstance {
-    config: NodeJS.Dict<string>,
-    addRawCustomPlugin<RequestType>(method: RawCustomPluginMethod, path: string, handler: AsyncHandler<RequestType> | Handler<RequestType>, schema?: InputOutputSchemas, advancedConfigs?: RawCustomPluginAdvancedConfig): DecoratedFastify,
-    addPreDecorator<RequestType>(path: string, handler: preDecoratorHandler<RequestType>): DecoratedFastify
-    addPostDecorator<RequestType>(path: string, handler: postDecoratorHandler<RequestType>): DecoratedFastify
+  interface DecoratedFastify<ServiceConfig = NodeJS.Dict<string>> extends fastify.FastifyInstance {
+    config: ServiceConfig,
+    addRawCustomPlugin<RouteGeneric extends RouteGenericInterface = RouteGenericInterface>(method: RawCustomPluginMethod, path: string, handler: AsyncHandler<ServiceConfig, RouteGeneric> | Handler<ServiceConfig, RouteGeneric>, schema?: InputOutputSchemas, advancedConfigs?: RawCustomPluginAdvancedConfig): DecoratedFastify,
```
Comment (ilteoood) [inline index.d.ts:63]: Same thing here

File: tests/types/index.ts [hunk 3]
```diff
@@ -145,7 +145,7 @@ a(async function (service) {
     headers: {
       type:'object'
     }
-  }).addRawCustomPlugin<{rawPlugin: string}>('GET', '/ts', async function handler(request, reply) {
+  }).addRawCustomPlugin<{Body: {rawPlugin: string}}>('GET', '/ts', async function handler(request, reply) {
```
Comment (ilteoood) [inline tests/types/index.ts:148]: I think that this type should be extracted, not inline

File: tests/types/index.ts [hunk 4]
```diff
@@ -182,7 +182,7 @@ a(async function (service) {
   service.addPreDecorator('/decorators/my-pre4', async function myHandlerPreDecorator(request, reply) {
     return request.abortChain(200, { final: 'body' }, { some: 'other headers' })
   })
-  service.addPreDecorator<{decorator: string}>('/decorators/my-pre5', async function myHandlerPreDecorator(request, response) {
+  service.addPreDecorator<{Body: {decorator: string}}>('/decorators/my-pre5', async function myHandlerPreDecorator(request, response) {
```
Comment (ilteoood) [inline tests/types/index.ts:185]: Same here

File: index.d.ts [hunk 5]
```diff
@@ -18,23 +18,24 @@ import * as fastify from 'fastify'
 import * as http from 'http'
 
 import {FormatName} from 'ajv-formats'
+import {RouteGenericInterface} from 'fastify/types/route'
```
Comment (davidebianchi) [inline index.d.ts:21]: This is not exported from fastify? It is possible to import only as a sub file?

File: tests/types/index.ts (hunk 4 above)
Comment (ilteoood) [reply tests/types/index.ts:185]: Can you check the others?

File: index.d.ts [hunk 6]
```diff
@@ -88,9 +89,19 @@ declare namespace customPlugin {
   //
   // CUSTOM PLUGIN
   //
-  type BasicHandler<RequestType, ResponseType> = (this: DecoratedFastify, request: DecoratedRequest<RequestType>, reply: fastify.FastifyReply) => ResponseType
-  type Handler<RequestType> = BasicHandler<RequestType, void>
-  type AsyncHandler<RequestType> = BasicHandler<RequestType, Promise<any>>
+  type BasicHandler<
+    ServiceConfigType = NodeJS.Dict<string>,
+    RouteGeneric extends RouteGenericInterface = RouteGenericInterface,
```
Comment (davidebianchi) [inline index.d.ts:94]: Those types are repeated often, it is not possible to assign to a specific type and reuse it?

**Review by davidebianchi — CHANGES_REQUESTED** [review]
Thanks for the changes, a few comments

File: index.d.ts (hunk 1 above)
Comment (epessina) [reply index.d.ts:38]: I've looked it up, but I can't find a way to extract a generic. It is not actually a type itself. 
Do you know how to do it?

File: index.d.ts (hunk 6 above)
Comment (epessina) [reply index.d.ts]: I've looked it up, but I can't find a way to extract a generic. It is not actually a type itself.
Do you know how to do it?

**Comment by epessina** [comment]
Ping

File: index.d.ts (hunk 1 above)
Comment (ilteoood) [reply index.d.ts:37]: What about something like:

```
type ServiceConfig<T = NodeJS.Dict<string>> = T
```

?

**Review by ilteoood — CHANGES_REQUESTED** [review]

**Review by davidebianchi — APPROVED** [review]

**Review by ilteoood — APPROVED** [review]
LGTM

**Pull request merged**